airflow/models/dagrun.py

```diff
@@ -966,6 +969,14 @@ def _filter_tis_and_exclude_removed(dag: DAG, tis: list[TI]) -> Iterable[TI]:
                 new_unfinished_tis = [t for t in unfinished_tis if t.state in State.unfinished]
                 finished_tis.extend(t for t in unfinished_tis if t.state in State.finished)
                 unfinished_tis = new_unfinished_tis
+
+            return schedulable_tis, changed_tis, unfinished_tis
+
+        if unfinished_tis:
+            schedulable_tis, changed_tis, unfinished_tis = _decide_on_tis(unfinished_tis)
+        elif not unfinished_tis and finished_tis:
+            schedulable_tis, changed_tis, unfinished_tis = _decide_on_tis(blocked_tis)
+            unfinished_tis = [t for t in unfinished_tis if t.state in State.unfinished]
```

Comment:
I'm a little unsure about this approach.

- I think we can simplify this and include blocked tasks if there are no unfinished/non-blocked tasks, instead of needing a helper function.
- It feels like waiting to unblock tasks until there are no remaining non-blocked tasks might be inefficient?

Reply:
Yeah, the helper function is not needed really. 

> It feels like waiting to unblock tasks until there are no remaining non-blocked tasks might be inefficient?

I don't think it will be inefficient considering the scheduler loop would be fast